### ISSUE TITLE
feature: Tidy 5e Rewrite Compatibility

### ISSUE THREAD

**Issue by kgar** [MrPrimate/tokenizer#148]
I am the author of the Tidy 5e Sheets rewrite. The rewritten sheets run on svelte and have some significant differences from the default sheets and even original Tidy. With that in mind, I am requesting to work with you on a way to restore compatibility with Tokenizer.

## Details

At present, the incompatibility is with my actor sheet portraits. They are using `mousedown` instead of click, and they do not have the `[data-edit="img"]` attribute in place. These are implementation details of my rewrite, but as an exercise, I lined things up as expected to see how far that would get me. I found that, because it's wired with svelte and not jQuery, `$(element).off("click")` doesn't affect my event bindings. As a result, shift+clicking causes 2 file pickers to appear: my original one and the one from the `else` case when the module decides whether to launch tokenizer.

One idea I have is to use a `pre` hook:
Like dnd5e's `preUseItem` hook, I could provide a hook that essentially provides the ability to hook into and prevent the default behavior when someone interacts with a sheet portrait, one hook for actors and one for items. Something like `tidy5e-sheet.preOpenActorPortraitFilePicker` with params `(actor, event, ...)`. I'm not 100% on the name or the params and am open to any suggestions.

Another idea I have is to use a command registration API:
I present this as an extra option in addition to or in place of other options, in case you are interested.
My portraits have additional options shown as buttons when right-clicking the portrait. I could add an API function for registering commands of your own to the portrait commands. In this case, registering "Open Tokenizer" or just "Tokenizer" as an option that appears to the user. Like any of my command registration API functions ([example](https://kgar.github.io/foundry-vtt-tidy-5e-sheets/classes/ItemSummaryApi.html#registerCommands)), this one would feature optional callbacks for when someone executes the command and whether it should be enabled. While they are rendered as buttons right now, they may eventually become context menu options, and a command registration API would ensure that the registered content is data-driven so that I can seamlessly change the button to a context menu option without you having to make any changes on your side.

---

In any case, I have created an inconvenience because of going off the beaten path in Foundry sheets development, but I would like to collaborate to create intentional compatibility between us.

I hope to hear your thoughts. Thanks for reading.

**Comment by MrPrimate** [comment]
I'm happy with either of the approaches - the functions that the sheet wrappers call are already exposed via an API, so there is always the fallback of that if either of the other two don't work. I would probably lean towards the api registration - that would work well, and lends itself well to having multiple tools able to bind themselves to the portrait etc

**Comment by kgar** [comment]
At last, I will be able to start working on this. 
I will provide hooks/API for both options. Please feel free to use either or both approaches as you see fit.

Thank you for working with me on this!

**Comment by kgar** [comment]
Here is a near-final draft of portrait menu command registration:
```js
Hooks.once("tidy5e-sheet.ready", (api) => {
  api.actorPortrait.registerMenuCommands([
    {
      label: "Test",
      iconClass: "fa-solid fa-flask",
      tooltip: "Click for test result",
      enabled: (params) => params.actor.type !== "vehicle",
      execute: (params) => {
        console.log(params);
        ui.notifications.info(
          "Hello, Test Portrait Menu Command for " + params.actor.name
        );
      },
    },
  ]);
});

```

![image](https://github.com/MrPrimate/tokenizer/assets/4632129/1b8abe39-194d-4993-9c19-a26d1f045f00)

![image](https://github.com/MrPrimate/tokenizer/assets/4632129/9fc563bb-13ff-4047-b0c8-881bf9600703)

![image](https://github.com/MrPrimate/tokenizer/assets/4632129/9a51dcdb-f3d4-4348-9b3c-19f90f08325b)

**Comment by kgar** [comment]
Here is the hook in action:
```js
Hooks.on('tidy5e-sheet.preOpenActorPortraitFilePicker', (actor, event) => {
  // Do something else
  return false; // prevent the portrait file picker from opening
});
```

This is set up to trigger for all actors when the user left-clicks / taps on the actor portrait.

I'm packaging up the release for this and will post it here when ready.

**Comment by kgar** [comment]
Tidy v0.2.9 has been released with the new [API function](<https://kgar.github.io/foundry-vtt-tidy-5e-sheets/classes/ActorPortraitApi.html#registerMenuCommands>) and the hook described above.

**Comment by MrPrimate** [comment]
On the `registerMenuCommands` `execute` `params` can you include the `token` data like the sheet render hook does if opened from a scene? This allows me to update the tokens on the current scene with the new token image, which is often where people are opening character sheets from.

**Comment by kgar** [comment]
What is the prop path on that usually? Is it like `data.options.token`?

**Comment by kgar** [comment]
Update; I have released v0.2.11 with the actor sheet `context` prop added to the `registerMenuCommands` `execute` `params` field.

https://kgar.github.io/foundry-vtt-tidy-5e-sheets/interfaces/PortraitContextMenuCommandExecuteParams.html#context

**Comment by MrPrimate** [comment]
Thank you, and out in v4.2.12

**Comment by kgar** [comment]
Thank you again for this!

**Comment by kgar** [comment]
Ah, one last thing: I realized I did not update the `tidy5e-sheet.preOpenActorPortraitFilePicker` with the `context` field, which prevents you from taking the original approach of handling clicking / shift+clicking the portrait.

I am updating that hook call to include the full context.

So, if a user posts an issue requesting the original left-click behavior on the portrait to open Tokenizer, the `pre` hook will include the `context` as of my next release in a day or two:

![image](https://github.com/MrPrimate/tokenizer/assets/4632129/47382eca-5a44-4182-bd89-ff7e1d7f3bf5)

This is just in case.

And thanks again! 🍻
